cluster/gce/trusty/node.yaml

```diff
@@ -148,59 +150,83 @@ post-stop exec sleep 10
 	. /etc/kube-env
-	ARGS="--v=2"
-	if [ -n "${KUBEPROXY_TEST_ARGS:-}" ]; then
-		ARGS="${KUBEPROXY_TEST_ARGS}"
-	fi
-	BINARY_PATH="/usr/bin/kube-proxy"
-	if [ "${TEST_CLUSTER:-}" = "true" ]; then
-		BINARY_PATH="/usr/local/bin/kube-proxy"
-	fi
-	${BINARY_PATH} \
-		--master=https://${KUBERNETES_MASTER_NAME} \
-		--kubeconfig=/var/lib/kube-proxy/kubeconfig \
-		${ARGS}
+	restart_docker_daemon
 end script
 
-# Wait for 10s to start kube-proxy again.
-post-stop exec sleep 10
-
 --====================================
 MIME-Version: 1.0
 Content-Type: text/upstart-job; charset="us-ascii"
 Content-Transfer-Encoding: 7bit
-Content-Disposition: attachment; filename="kube-docker.conf"
+Content-Disposition: attachment; filename="kube-proxy.conf"
 
 #upstart-job
 
-description "Restart docker daemon"
+description "Start kube-proxy static pod"
 
-# The condition "stopped kube-install-additional-packages" is to avoid
-# breaking nsenter installation, which is through a docker container.
-# It can be removed if we find a better way to install nsenter.
-start on started kubelet and stopped kube-install-additional-packages
+start on stopped kube-docker
 
 script
 	set -o errexit
 	set -o nounset
 
 	. /etc/kube-configure.sh
-	restart_docker_daemon
+	. /etc/kube-env
+	prepare_log_file "/var/log/kube-proxy.log"
+	# Load the docker image from file /run/kube-docker-files/kube-proxy.tar.
+	for i in {1..3}; do
+		echo "Try to load docker image file kube-proxy.tar"
+		timeout 30 docker load -i /run/kube-docker-files/kube-proxy.tar 1>/dev/null 2>&1
+		if [ "$?" = "0" ]; then
+			break
+		fi
+		sleep 5
+	done
+	# Copy the manifest to /tmp to manipulate
+	tmp_file="/tmp/kube-proxy.manifest"
+	cp -f /run/kube-manifests/kubernetes/kube-proxy.manifest ${tmp_file}
+	# Remove the lines of salt configuration and replace variables with values.
+	# NOTE: Changes to variable names in cluster/saltbase/salt/kube-proxy/kube-proxy.manifest
+	# may break this upstart job.
+	sed -i "/^ *{%/d" ${tmp_file}
+	kubeconfig="--kubeconfig=\/var\/lib\/kube-proxy\/kubeconfig"
+	kube_docker_registry="gcr.io\/google_containers"
+	if [ -n "${KUBE_DOCKER_REGISTRY:-}" ]; then
+		kube_docker_registry=${KUBE_DOCKER_REGISTRY}
+	fi
+	kube_proxy_docker_tag=$(cat /run/kube-docker-files/kube-proxy.docker_tag)
+	test_args=""
+	log_level="--v=2"
+	if [ -n "${KUBEPROXY_TEST_ARGS:-}" ]; then
+		test_args="${KUBEPROXY_TEST_ARGS}"
+		# test_args should already contain log level setting.
+		log_level=""
+	fi
+	api_servers="--master=https:\/\/${KUBERNETES_MASTER_NAME}"
+	sed -i -e "s/{{kubeconfig}}/${kubeconfig}/g" ${tmp_file}
+	sed -i -e "s/{{pillar\['kube_docker_registry'\]}}/${kube_docker_registry}/g" ${tmp_file}
+	sed -i -e "s/{{pillar\['kube-proxy_docker_tag'\]}}/${kube_proxy_docker_tag}/g" ${tmp_file}
+	sed -i -e "s/{{test_args}}/${test_args}/g" ${tmp_file}
+	sed -i -e "s/{{log_level}}/${log_level}/g" ${tmp_file}
+	sed -i -e "s/{{api_servers_with_port}}/${api_servers}/g" ${tmp_file}
+
```

Comment:
Frankly speaking, I am not convinced. As I said, making copies will make you monitor every change in salt files and make changes in your copies. If you do this for every manifest with salt content, it will be a big amount of effort to maintain the copies, especially considering that merging PRs is slow in many cases. It is not a clear logic to say why not just use salt. We do not use it because we don't want to have such a complex thing in our trusty stuff. I am not sure what you meant "special cases". The shell code is super simple, just several lines of "sed" commands. I don't think it is a burden to maintain it, comparing with monitoring every change in salt and then making corresponding changes in manifests copies. Two options: (1) maintaining N manifests + 1 shell file once per week; (2) maintaining N manifests 3 times per week. I would like to go with the option (1), unless there is a strong reason for (2).

Another reason I try to avoid static copies: We have already set up Jenkins auto tests running against Trusty. Using static copies is not helpful for catching problems. You can imagine, someone updates the salt file but I do not update the copy, the tests will still pass probably. But if using the file from salt, a breakage will make kube-proxy fail to start or behave abnormally, then auto tests can catch it.